### ISSUE TITLE
Proposal: Create helm/tiller operator

### ISSUE THREAD

**Issue by burdiyan** [kubernetes/helm#3089]
Right now there is no easy and blessed way for configuring a set of Helm releases to be deployed in Kubernetes cluster. There is no way to express the desired state of a release.

Although there are tools like https://github.com/roboll/helmfile and https://github.com/eneco/landscaper that allow to achieve something similar, these are not ideal.

I propose to create an Operator for Helm/Tiller that runs in the cluster and can manage custom resources that would describe Helm releases.

This would allow to describe the desired state (the thing we used to so much with Kubernetes) for Helm releases.

Users could submit these resources with plain `kubectl create/apply` and Operator would rollout the desired release across the cluster.

I can see a structure similar to something like:

```
apiVersion: helm.sh/v1
kind: Release
metadata:
  name: nginx-ingress
  namespace: kube-public
spec:
  chart: stable/nginx-ingress
  reuseValues: true
  values:
    rbac:
      create: true
...
```

There are a lot of things to take into account, but I guess this will be helpful for many people.

**Comment by burdiyan** [comment]
What I think also great about this approach, it would be very easy to run a chart repository inside the cluster, that will be not exposed to the world. So Kubernetes users don't need to have access to the chart repository. They only would need access to create normal Kubernetes resources, and the Operator is the one who is authenticated agains the chart repository.

**Comment by bacongobbler** [comment]
There was a similar discussion raised in the [application definition working group](https://github.com/kubernetes/community/tree/master/wg-app-def). If you have ideas, I'd suggest joining that working group!

**Comment by bacongobbler** [comment]
Also for what it's worth, this is exactly how tiller works today. The only difference is that tiller also acts as the templating engine and spits out configmaps containing release information, after which tiller reconciles the deployment.

The only thing imperative about the workflow is the helm CLI itself, which waits for the deployment status to change from DEPLOYING to SUCCESS/FAILURE depending on the release's deployment status. It would be quite simple for someone to write a helm-like tool that communicates directly to tiller and performs the deployment in a declarative fashion, if my assumptions are correct.

**Comment by technosophos** [comment]
I presume that by Operator you mean Controller?

There was a major effort to implement this about a year ago, and it ultimately failed for a wide number of reasons. We agreed to re-visit it during Helm 3 planning. As @bacongobbler points out, the relevant starting point is #1650. Tamal's PR there shows how far they got, and why it ended up not working.

**Comment by burdiyan** [comment]
@technosophos I was meaning Operator in a sense of CoreOS' Operator concept (https://coreos.com/operators/), which is indeed a Controller essentially. 

I'll take a look at Tamal's PR. Thanks a lot!

**Comment by burdiyan** [comment]
As far as I understood, the mentioned PR was trying to implement storage backend for Tiller using Third Party Resources in Kubernetes (which is something I always was thinking about).

Maybe instead of implementing another storage backend, this Operator/Controller could just watch the custom resource, and then interact with Tiller as if it were Helm CLI. This way no changes for Helm/Tiller would be needed, and ConfigMap can continue be the default storage for Tiller.

That seems like a workaround, but until Helm 3.0 this would probably be the only solution.

@technosophos Do you already have any clear plans about Helm 3.0?

**Comment by puja108** [comment]
We'd be interested in this, too. The operator concept sounds quite fitting for this.

Currently, we are using an in-cluster agent that implements pull-based-deployments (https://github.com/giantswarm/draughtsman) for us talking to Tiller, but an operator could help make this more reliable, especially when I hear you talking about reconcilliation.

We were thinking of making it an operator (based on https://github.com/giantswarm/operatorkit/), but there's no priority on that at the moment.

**Comment by technosophos** [comment]
Yes, these ideas have been proposed. And we are going to talk through them at the Helm Summit (You all should come!). That event will be where we formally launch the Helm 3 design process.

I'm still not sure what an operator would buy, in this model, over and above a controller. But I'd entertain the idea. We had designed a version that had a CRD controller in front of a Tiller proxy (@adamreese whiteboarded it out with me), but it did not solve our auth problem at the time. We could not assert anything about the identity of the submitter, and consequently could not establish trust. Essentially, anyone who could create a CRD could do an install, but we could not tell who that user was. With just Tiller, you can at least lock things down with namespaces and RBACS, though ultimately it suffers from a similar problem with identity. Hopefully the Pod Identity WG will solve the necessary bits of that and we can move along that path.

In short, though, there's probably not a lot of utility to be had in trying to add CRD support to Helm 2 when Helm 3 work is just around the corner. It would be better to start working on a Helm 3 proposal.

So the main question would be: If you only had to keep the chart, and the basic release process we have now (e.g. install/upgrade/rollback/delete/show), how would you do it? What would the process look like end-to-end? And how would you handle things like `--dry-run`?

If CRDs (or even TPRs) had been around/stable when we originally designed Helm in late 2015, I suspect the system would be quite a bit different than it is now.

**Comment by nicolerenee** [comment]
@burdiyan my team and I have build an open source project that might do exactly what you need. Take a look it's at https://github.com/wpengine/lostromos. I'd be willing to answer any questions you may have with setting it up or helping you use it for your use case.

**Comment by anguslees** [comment]
I wrote one last weekend fwiw: https://github.com/bitnami/helm-crd
File issues/PRs on that repo if you encounter any issues with it.

**Comment by mcwienczek** [comment]
There are a few days to approach this:
1. Implement MVP in a form of minimal feature set required to successfully use the Operator
2. Start working on it, but announce to the community that it is an alpha feature. This is exactly the Kubernetes approach.
3. Take something like @anguslees or @nicolerenee have created and incorporate it into helm source code

First though, I think we need Technical Design Document for that so we have any idea how this might work and look like. 

I would like to help with both implementation and design, but it definitely requires strong buy-in from core maintainers first.

@technosophos where can I find information about Helm Summit? I probably won't be able to come anyway as I live in EU, but still :) .

**Comment by bacongobbler** [comment]
We're still ironing out some things in prep for the summit, but it'll likely occur in Portland in early 2018. Stay tuned in #helm-users!

**Comment by nicolerenee** [comment]
I would like to be involved in building out the Technical Design Document as well as helping with the implementation.

**Comment by burdiyan** [comment]
We finally revisited our workflow, and since we managing our infrastructure with Terraform we managed to handle Helm releases with Terraform in a declarative way using a custom plugin.

I also developed a plugin for Helm that allows to change values of a running release without specifying the chart again: https://github.com/burdiyan/helm-update-config/.

So we initially deploy all the charts with Terraform, and then each of our services on CI just need to change the Docker image tag with helm-update-config to deploy a new version of itself.

Anyway, I still really believe that the subject of this ticket can be pretty helpful for other people and worth being implemented!

**Comment by kfox1111** [comment]
helm-update-config looks super helpful. Thanks. :)

I have to admit, I've accidentally updated the package along with the config on a couple of occasions. This seems like a good safety net.

Can this be rolled into helm cli natively?

**Comment by burdiyan** [comment]
Thank!

I’ll be happy to submit a PR if maintainers are seeing it useful.

**Comment by burdiyan** [comment]
Thanks!

I’ll be happy to submit a PR if maintainers are seeing it useful.

**Comment by puja108** [comment]
As an update from our side, we started work on https://github.com/giantswarm/chart-operator which will be an operator that reconciles Helm charts against a CNR (should work with self-hosted and Quay.io). Feedback on that approach more than welcome.

**Comment by saulshanabrook** [comment]
> So we initially deploy all the charts with Terraform, and then each of our services on CI just need to change the Docker image tag with helm-update-config to deploy a new version of itself.

@burdiyan Do you use this provider to provision the helm charts with Terraform? https://github.com/mcuadros/terraform-provider-helm

**Comment by bacongobbler** [comment]
I also would like to point out that this is noted in the helm 3 proposal via https://github.com/kubernetes-helm/community/blob/master/helm-v3/008-controller.md

Basically the proposal boils down to this (from the controller's point of view):

> A client-only Helm follows a push-based model. To facilitate a pull-based model, we can also create a Helm controller that provides a CRD facade, but uses the Helm library.

> It is suggested that this (a) be a core Helm project, but (b) is kept separately from the main Helm codebase.

If I'm understanding the proposal correctly, this would likely be a community-owned project at https://github.com/kubernetes-helm/.

I'll let one of the other core maintainers (@technosophos?) chime in here if they have any thoughts. I'm just relaying what I know. :)

**Comment by davidkarlsen** [comment]
Also see https://github.com/weaveworks/flux-helm-test which does this.
